system/lib/standalone/standalone.c

```diff
@@ -65,53 +81,1009 @@ weak int _munmap_js(
+      // The close-on-exec flag is ignored.
+      return 0;
+    case F_GETFL: {
+      // Obtain the flags and the rights of the descriptor.
+      __wasi_fdstat_t fds;
+      __wasi_errno_t error = __wasi_fd_fdstat_get(fd, &fds);
+      if (error != 0) {
+        errno = error;
+        return -1;
+      }
+
+      // Roughly approximate the access mode by converting the rights.
+      int oflags = fds.fs_flags;
+      if ((fds.fs_rights_base &
+           (__WASI_RIGHTS_FD_READ | __WASI_RIGHTS_FD_READDIR)) != 0) {
+        if ((fds.fs_rights_base & __WASI_RIGHTS_FD_WRITE) != 0)
+          oflags |= O_RDWR;
+        else
+          oflags |= O_RDONLY;
+      } else if ((fds.fs_rights_base & __WASI_RIGHTS_FD_WRITE) != 0) {
+        oflags |= O_WRONLY;
+      } else {
+        oflags |= O_SEARCH;
+      }
+      return oflags;
+    }
+    case F_SETFL: {
+      // Set new file descriptor flags.
+      va_list ap;
+      va_start(ap, cmd);
+      int flags = va_arg(ap, int);
+      va_end(ap);
+
+      __wasi_fdflags_t fs_flags = flags & 0xfff;
+      __wasi_errno_t error =
+        __wasi_fd_fdstat_set_flags(fd, fs_flags);
+      if (error != 0) {
+        errno = error;
+        return -1;
+      }
+      return 0;
+    }
+    default:
+      errno = EINVAL;
+      return -1;
+  }
 }
 
-weak int __syscall_fstat64(int fd, intptr_t buf) {
-  return -ENOSYS;
+weak int __syscall_ftruncate64(int fd, off_t length) {
+  if (length < 0) {
+    errno = EINVAL;
+    return -1;
+  }
+  __wasi_filesize_t st_size = length;
+  __wasi_errno_t error =
+    __wasi_fd_filestat_set_size(fd, st_size);
+  if (error != 0) {
+    errno = error;
+    return -1;
+  }
+  return 0;
 }
 
-weak int __syscall_stat64(intptr_t path, intptr_t buf) {
+int rmdirat(int dirfd, intptr_t path) {
+  const char* resolved_path = (const char*)path;
+
+  // Resolve path if fd is AT_FDCWD.
+  if (dirfd == AT_FDCWD) {
+    char *relative_path;
+    dirfd = find_relpath(resolved_path, &relative_path);
+
+    // If we can't find a preopen for it, fail as if we can't find the path.
+    if (dirfd == -1) {
+      errno = ENOENT;
+      return -1;
+    }
+
+    resolved_path = relative_path;
+  }
+
+  __wasi_errno_t error = __wasi_path_remove_directory(dirfd, resolved_path, strlen(resolved_path));
+  if (error != 0) {
+    errno = error;
+    return -1;
+  }
+
+  return 0;
+}
+
+weak int __syscall_rmdir(intptr_t path) {
+  return rmdirat(AT_FDCWD, path);
+}
+
+weak int __syscall_unlinkat(int dirfd, intptr_t path, int flags) {
+  // unlinkat with AT_REMOVEDIR flag is acutally rmdir.
+  if ((flags & AT_REMOVEDIR) != 0) {
+    return rmdirat(dirfd, path);
+  }
+
+  const char* resolved_path = (const char*)path;
+
+  // Resolve path if fd is AT_FDCWD.
+  if (dirfd == AT_FDCWD) {
+    char *relative_path;
+    dirfd = find_relpath(resolved_path, &relative_path);
+
+    // If we can't find a preopen for it, fail as if we can't find the path.
+    if (dirfd == -1) {
+      errno = ENOENT;
```

Comment:
I think this should be `return -ENOENT`. And similar elsewhere. In wasi-libc, this code is used in function that use the "return -1 + errno" error convention, while the `__syscall_*` functions use the "return negative error" convention of the Linux syscall API.